File: MY_REPOS/Lambda-Resource-Static-Assets/2-content/14-Pure-Education/java-data-struc/java-prep/src/main/java/data/structures/java/trees/Tree.java
package data.structures.java.trees;

import java.util.Stack;

public class Tree<T extends Comparable<T>>
{
  public TreeNode<T> root;

  public Tree(T[] values)
  {
    for(T val : values)
    {
      add(val);
    }
  }

  void add(T value)
  {
    if(root == null)
    {
      root = new TreeNode<>(value);
    }
    else
    {
      add(root, value);
    }
  }

  private void add(TreeNode<T> treeNode, T value)
  {
    if(value.compareTo(treeNode.data) == -1)
    {
      if(treeNode.left == null)
      {
        treeNode.left = new TreeNode<>(value);
      }
      else
      {
        add(treeNode.left, value);
      }
    }
    else
    {
      if(treeNode.right == null)
      {
        treeNode.right = new TreeNode<>(value);
      }
      else
      {
        add(treeNode.right, value);
      }
    }
  }

  Iterator iterator()
  {
    return new Iterator(root);
  }

  class Iterator
  {
    private Stack<TreeNode<T>> stack = new Stack<>();

    Iterator(TreeNode<T> treeNode)
    {
      while(treeNode != null)
      {
        stack.push(treeNode);
        treeNode = treeNode.left;
      }
    }

    boolean hasNext()
    {
      return stack.empty();
    }

    TreeNode<T> next()
    {
      TreeNode<T> treeNode = stack.pop();

      if(treeNode.right != null)
      {
        TreeNode<T> tmp = treeNode.right;
        while(tmp != null)
        {
          stack.push(tmp);
          tmp = tmp.left;
        }
      }
      return treeNode;
    }
  }
}

File: MY_REPOS/DS-AND-ALGO-Notes-P2/notes/D1/my/tree.js
class TreeNode {
  constructor(val) {
    this.val = val;
    this.left = null;
    this.right = null;
  }
}
let a = new TreeNode("a");
let b = new TreeNode("b");
let c = new TreeNode("c");
let d = new TreeNode("d");
let e = new TreeNode("e");
let f = new TreeNode("f");

a.right = c;
b.left = d;
b.right = e;
c.right = f;
console.log(a);
/*
  TreeNode {
  val: 'a',
  left: null,
  right: TreeNode {
    val: 'c',
    left: null,
    right: TreeNode { val: 'f', left: null, right: null }
  }
}
*/
function inOrderPrint(root) {
  if ((root = null)) return; //base case
  inOrderPrint(root.left);
  console.log(root.val);
  inOrderPrint(root.right);
}

//!
